### PR TITLE
introduce Task and TaskRunner; record build status to Firebase

### PR DESCRIPTION
**Pull request by yjbanov**
This sets up infrastructure to fix the multiple dashboard issues.

/cc @devoncarew

### CODE REVIEW

File: bin/build.dart [hunk 1]
```diff
@@ -59,17 +61,24 @@ Future<Null> build() async {
...
+    ..enqueueAll(createPerfTests())
+    ..enqueueAll(createStartupTests())
+    ..enqueue(createGalleryTest())
+    ..enqueueAll(createAnalyzerTests(sdk: sdk, commit: commit, timestamp: timestamp))
+    ..enqueue(createRefreshTest());
+
+  BuildResult result = await runner.run();
+  section('Build results');
+  print('Ran ${result.results.length} tasks (${result.failedTaskCount} failed):');
+  for (TaskResult taskResult in result.results) {
```
Comment (devoncarew) [inline bin/build.dart]: nit: no braces needed

File: lib/src/analysis.dart [hunk 2]
```diff
@@ -8,20 +8,32 @@ import 'dart:io';
 import 'package:path/path.dart' as path;
 
 import 'benchmarks.dart';
+import 'framework.dart';
 import 'utils.dart';
 
-Future<Null> runAnalyzerTests({
+List<Task> createAnalyzerTests({
```
Comment (devoncarew) [inline lib/src/analysis.dart]: createAnalyzerTests ==> `createAnalyzerTasks`?

File: lib/src/perf_tests.dart [hunk 3]
```diff
@@ -0,0 +1,65 @@
+// Copyright (c) 2016 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+
+import 'framework.dart';
+import 'utils.dart';
+
+List<Task> createPerfTests() {
```
Comment (devoncarew) [inline lib/src/perf_tests.dart]: same, do you want to call this `createPerfTests` or `createPerfTasks`?

File: lib/src/framework.dart [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+  void enqueueAll(Iterable<Task> tasks) {
+    _taskQueue.addAll(tasks);
+  }
+
+  Future<BuildResult> run() async {
+    List<TaskResult> results = <TaskResult>[];
+
+    for (Task task in _taskQueue) {
+      section('Running task ${task.name}');
```
Comment (devoncarew) [inline lib/src/framework.dart]: I would ensure that the individual tasks don't also print out sections at the start, just so we don't get duplicated printouts.

**Comment by devoncarew** [comment]
awesome, lgtm

File: bin/build.dart (hunk 1 above)
Comment (yjbanov) [inline bin/build.dart]: Done.

File: lib/src/analysis.dart (hunk 2 above)
Comment (yjbanov) [inline lib/src/analysis.dart]: Not sure putting the type in the name is idiomatic. You wouldn't write `List<int> runtimeInts` or `List<String> nameStrings`, but `List<int> runtimes` and `List<String> names`.

File: lib/src/framework.dart (hunk 4 above)
Comment (yjbanov) [inline lib/src/framework.dart]: Didn't find any. Should be fine.

**Pull request merged**